In [ ]:
import pandas as pd
pd.set_option("max.colwidth",0)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Data Information

<b>Context</b>
- This dataset is originally from the National Institute of Diabetes and Digestive and Kidney Diseases. The objective is to predict based on diagnostic measurements whether a patient has diabetes.

<b>Content</b>
- Several constraints were placed on the selection of these instances from a larger database. In particular, all patients here are females at least 21 years old of Pima Indian heritage.

<b>Data Attributes</b>
- Pregnancies: Number of times pregnant
- Glucose: Plasma glucose concentration a 2 hours in an oral glucose tolerance test
- BloodPressure: Diastolic blood pressure (mm Hg)
- SkinThickness: Triceps skin fold thickness (mm)
- Insulin: 2-Hour serum insulin (mu U/ml)
- BMI: Body mass index (weight in kg/(height in m)^2)
- DiabetesPedigreeFunction: Diabetes pedigree function
- Age: Age (years)
- Outcome: Class variable (0 or 1)

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/pandas/Copy of diabetes.csv")


In [ ]:
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0  6            148      72             35             0        33.6   
1  1            85       66             29             0        26.6   
2  8            183      64             0              0        23.3   
3  1            89       66             23             94       28.1   
4  0            137      40             35             168      43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0  0.627                     50   1        
1  0.351                     31   0        
2  0.672                     32   1        
3  0.167                     21   0        
4  2.288                     33   1

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


# Step 1: Define Business Questions — Think Like a Data Analyst, Not Just a Coder
Before touching the data, `what questions` actually matter? Here are some examples to get them started, but they should come up with your own ideas:

- Which features (e.g., glucose level, BMI) are strongest predictors of diabetes?

- Are there notable differences in average values of health indicators between diabetic and non-diabetic groups?

- Does age or number of pregnancies influence diabetes likelihood?

- What health factors could a medical clinic focus on to identify at-risk patients early?

- Can we segment patients into meaningful risk groups for targeted interventions?

Make it clear: Without a business or health context, data analysis can be pointless or misleading.

# Step 2: Explore the Dataset with Pandas
You should:

- Load the data into a DataFrame.

- Check basic info: shape, data types, missing or zero values that may require cleaning or imputation.

- Compute summary stats (mean, median, std) grouped by Outcome.

- Correlation matrix to find relationships.

- Maybe simple pivot tables to cross-examine features like Age vs. Outcome.

# Step 3: Write the Analytical Report — Tell a Story, Not Just Numbers

- Objective: What questions are you answering?

- Data Description: Summarize the dataset, any cleaning steps.

- Analysis: Show key findings with code snippets or charts.

- Insights: What do these findings mean for the business or healthcare decision-making? For example:

    - "Higher glucose levels strongly associate with diabetes outcome, so glucose screening is critical."

    - "Older age groups show higher diabetes incidence—suggesting targeted age-based preventive programs."

- Limitations & Next Steps: What can’t this data tell us? What additional data or analysis is needed?



# Sample

Business Questions :

<b>Which factors in this dataset are most strongly associated with the presence of diabetes, and how could a health clinic use this information to identify high-risk patients?<b>



In [ ]:
print(data.info())
print(data.describe())

zero_cols = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
for col in zero_cols:
    print(f'Zeros in {col}:', (data[col] == 0).sum())


grouped = data.groupby('Outcome').mean()
print(grouped)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None
       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count  768.000000   768.000000  768.000000     768.000000     768.000000   
mean   3.845052     120.894531  69.105469      20.536458      79.799479    
std    

In [ ]:
corr = data.corr()
print(corr['Outcome'].sort_values(ascending=False))

Outcome                     1.000000
Glucose                     0.466581
BMI                         0.292695
Age                         0.238356
Pregnancies                 0.221898
DiabetesPedigreeFunction    0.173844
Insulin                     0.130548
SkinThickness               0.074752
BloodPressure               0.065068
Name: Outcome, dtype: float64


Key Findings :

- Glucose has the highest positive correlation with diabetes outcome (~0.47).

- BMI, Age, and DiabetesPedigreeFunction also show moderate positive correlations.

- Several features have zero values that likely indicate missing data rather than true zero (e.g., Insulin, SkinThickness).

- Pregnancies have a weaker correlation, but could still be a factor when combined with others.



# Note on Using LLMs (ChatGPT, Claude, etc.) for This Exercise
Feel free to use large language models (LLMs) to help generate ideas on what business questions to ask, what analyses to try, or how to interpret your results. They can be excellent for:

- Suggesting relevant features or hypotheses you might not think of.

- Providing example code snippets or pandas commands.

- Offering ways to structure your report or explain results clearly.

BUT, here’s the reality check:
- <font color="red">LLMs don’t “know” your specific dataset. They can’t see your data, they don’t check your numbers, and they don’t debug your code.</font> Their advice is often generic or based on patterns learned from many datasets, but:

- They may suggest complicated techniques or models that don’t fit your data or problem well.

- They often don’t spot data quality issues like missing values disguised as zeros or subtle biases.

- They can generate plausible-sounding but inaccurate interpretations if you don’t verify.

<font color="red">Trust yourself more than the LLM. </font>
- Double-check every suggestion, run your own exploratory analysis, and critically evaluate any output or recommendation.
- The power to deliver real insights lies in your understanding of the data and domain; not in blindly copying AI outputs.


Most importantly:

- Sometimes, you might want to uncover certain business insights, but the data simply doesn’t contain that information or signal. <b>That’s a completely valid outcome!</b>

# Now It’s Your Turn
explore the dataset and bring forward at least 5 key business insights related to diabetes risk or health indicators

In [ ]:
import pandas as pd

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/pandas/Copy of diabetes.csv")

In [ ]:
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin   BMI  \
0            6    148.0           72.0       35.00000  155.548223  33.6   
1            1     85.0           66.0       29.00000  155.548223  26.6   
2            8    183.0           64.0       29.15342  155.548223  23.3   
3            1     89.0           66.0       23.00000   94.000000  28.1   
4            0    137.0           40.0       35.00000  168.000000  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
data.shape

(768, 9)

In [ ]:
data.dtypes

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [ ]:
data.describe().T

count        mean         std     min       25%  \
Pregnancies               768.0    3.845052    3.369578   0.000   1.00000   
Glucose                   768.0  120.894531   31.972618   0.000  99.00000   
BloodPressure             768.0   69.105469   19.355807   0.000  62.00000   
SkinThickness             768.0   20.536458   15.952218   0.000   0.00000   
Insulin                   768.0   79.799479  115.244002   0.000   0.00000   
BMI                       768.0   31.992578    7.884160   0.000  27.30000   
DiabetesPedigreeFunction  768.0    0.471876    0.331329   0.078   0.24375   
Age                       768.0   33.240885   11.760232  21.000  24.00000   
Outcome                   768.0    0.348958    0.476951   0.000   0.00000   

                               50%        75%     max  
Pregnancies                 3.0000    6.00000   17.00  
Glucose                   117.0000  140.25000  199.00  
BloodPressure              72.0000   80.00000  122.00  
SkinThickness              23.0000   32.00000   99.00  
Insulin                    30.5000  127.25000  846.00  
BMI                        32.0000   36.60000   67.10  
DiabetesPedigreeFunction    0.3725    0.62625    2.42  
Age                        29.0000   41.00000   81.00  
Outcome                     0.0000    1.00000    1.00

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
(data[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']]==0).sum()

Glucose            5
BloodPressure     35
SkinThickness    227
Insulin          374
BMI               11
dtype: int64

In [ ]:
# Changing 0 value to NaN

In [ ]:
import numpy as np

cols_with_missing = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']

for col in cols_with_missing:
    data[col] = data[col].replace(0, np.nan)

data.isna().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

In [ ]:
# filling NaN value with median

In [ ]:
cols_with_missing = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']

for col in cols_with_missing:
    data[col].fillna(data[col].mean(), inplace=True)

/tmp/ipython-input-2163451615.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data[col].fillna(data[col].mean(), inplace=True)


In [ ]:
data.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [ ]:
# Outcome

In [ ]:
data['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

In [ ]:
data.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  110.710121      70.935397      27.768651  142.210761   
1           4.865672  142.165573      75.147324      31.736944  180.431548   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.888434                  0.429734  31.190000  
1        35.384757                  0.550500  37.067164

I used descriptive statistics (mean, median, min, max, std) with the groupby method and a pivot table to compare average health indicators by age and diabetes outcome. Missing values were handled, and the analysis focused on trends between diabetic and non-diabetic individuals across different ages.


In [ ]:
grouped_age = data.groupby('Outcome')['Age'].agg(['mean', 'median', 'min', 'max','std'])
grouped_age

mean  median  min  max        std
Outcome                                        
0        31.190000    27.0   21   81  11.667655
1        37.067164    36.0   21   70  10.968254

# Analysis of Age
The statistics for Age show that it is a relevant factor in diabetes prediction.

**Age Range**: Both groups start at the same minimum age of **21**. However, the non-diabetic group(Outcome=0) has a higher maximum age of **81**, compared to **70** for the diabetic group(Outcome=1).

**Average Age**: The average age for the diabetic group(Outcome=1) is **37.06**, which is noticeably older than the **31.19** average for the non-diabetic group. This suggests that older individuals are more likely to have diabetes.

**Median**: The median age of diabetic individuals is higher **(36)** than non-diabetics **(27)**, indicating that diabetes is more common in older age groups.

**Standard Devitation(std)**: It is slightly similar for both groups. This shows that the variation in ages is comparable, but it does not strongly affect the main insight. What matters more is that the mean and median ages are higher for diabetic individuals, indicating that older people are more likely to have diabetes.



In [ ]:
grouped_insulin = data.groupby('Outcome')['Insulin'].agg(['mean', 'median', 'min', 'max','std'])
grouped_insulin

mean      median   min    max        std
Outcome                                                
0        142.210761  155.548223  15.0  744.0  75.463785
1        180.431548  155.548223  14.0  846.0  95.747538

# Analysis of Insulin
The statistics for **Insulin** show that it is a relevant factor in diabetes prediction.



In [ ]:
grouped_glucose = data.groupby('Outcome')['Glucose'].agg(['mean', 'median', 'min', 'max','std'])
grouped_glucose

mean  median   min    max       std
Outcome                                           
0        110.710121   107.5  44.0  197.0  24.71706
1        142.165573   140.0  78.0  199.0  29.54175

In [ ]:
grouped_pregnancy = data.groupby('Outcome')['Pregnancies'].agg(['mean', 'median', 'min', 'max','std'])
grouped_pregnancy

mean  median  min  max       std
Outcome                                      
0        3.298000     2.0    0   13  3.017185
1        4.865672     4.0    0   17  3.741239

In [ ]:
grouped_BMI = data.groupby('Outcome')['BMI'].agg(['mean', 'median', 'min', 'max','std'])
grouped_BMI

mean  median   min   max       std
Outcome                                         
0        30.888434   30.40  18.2  57.3  6.504779
1        35.384757   34.25  22.9  67.1  6.595065

In [ ]:
grouped_bloodpressure = data.groupby('Outcome')['BloodPressure'].agg(['mean', 'median', 'min', 'max','std'])
grouped_bloodpressure

mean  median   min    max        std
Outcome                                           
0        70.935397    72.0  24.0  122.0  11.931033
1        75.147324    74.0  30.0  114.0  11.945712

In [ ]:
grouped_DPF = data.groupby('Outcome')['DiabetesPedigreeFunction'].agg(['mean', 'median', 'min', 'max','std'])
grouped_DPF

mean  median    min    max       std
Outcome                                          
0        0.429734   0.336  0.078  2.329  0.299085
1        0.550500   0.449  0.088  2.420  0.372354

In [ ]:
grouped_skinthickness = data.groupby('Outcome')['SkinThickness'].agg(['mean', 'median', 'min', 'max','std'])
grouped_skinthickness

mean    median  min   max       std
Outcome                                          
0        27.768651  29.15342  7.0  60.0  8.559606
1        31.736944  29.15342  7.0  99.0  8.647599

Correlation with Outcome

In [ ]:
corr = data.corr()
print(corr['Outcome'].sort_values(ascending=False))

Outcome                     1.000000
Glucose                     0.492928
BMI                         0.311924
Age                         0.238356
Pregnancies                 0.221898
SkinThickness               0.215299
Insulin                     0.214411
DiabetesPedigreeFunction    0.173844
BloodPressure               0.166074
Name: Outcome, dtype: float64


In [ ]:
pivot_age = pd.pivot_table(data, index='Age', columns='Outcome', aggfunc='mean')
pivot_age

BMI            BloodPressure             \
Outcome          0          1             0          1   
Age                                                      
21       28.656421  37.560000     66.765607  70.800000   
22       29.575655  35.045455     65.724845  72.363636   
23       30.693548  35.085714     69.058399  74.343598   
24       32.117302  38.775000     67.958031  68.800648   
25       30.269337  38.328571     64.741634  68.028942   
26       32.682299  45.950000     68.016207  70.300648   
27       30.570833  36.087500     75.250000  77.300648   
28       32.788000  35.780000     73.416207  70.040518   
29       33.500000  33.592308     73.462824  72.954245   
30       31.080000  32.826244     73.600000  67.135061   
31       30.936364  36.623077     70.891852  75.569630   
32       30.314286  33.877778     71.142857  77.267243   
33       31.871429  32.660000     63.142857  67.400000   
34       30.860000  31.925000     77.000000  66.500000   
35       33.300000  34.260000     75.200000  76.000000   
36       26.900000  34.610000     73.333333  73.840518   
37       28.546154  39.733333     75.800399  88.333333   
38       34.716667  36.080000     81.000000  72.440518   
39       32.722222  29.766667     74.666667  66.666667   
40       30.271429  37.350000     73.428571  76.400864   
41       33.744444  36.307692     72.600576  75.261937   
42       35.427273  34.285714     71.545455  76.285714   
43       33.800000  37.454545     66.000000  80.727273   
44       32.633333  35.080000     78.135061  80.881037   
45       34.357143  35.487500     81.428571  84.500000   
46       32.400000  36.342857     77.666667  74.571429   
47       40.250000  31.725000     92.000000  71.500000   
48       26.900000  42.300000     78.000000  80.000000   
49       36.050000  29.333333     78.000000  83.666667   
50       32.700000  30.340000     85.333333  74.000000   
51       34.200000  33.840000     88.666667  82.000000   
52       34.200000  33.371429     70.000000  83.142857   
53       23.300000  32.300000     75.000000  80.000000   
54       35.200000  36.714366     88.000000  90.000000   
55       27.666667  25.100000     70.333333  70.000000   
56       23.100000  36.000000     72.000000  78.500000   
57       27.750000  37.500000     77.000000  76.000000   
58       30.525000  34.966667     73.500000  84.666667   
59       27.300000  26.800000     84.000000  69.000000   
60       29.933333  26.950000     86.666667  70.000000   
61       28.800000  31.200000     60.000000  92.000000   
62       26.600000  31.300000     66.000000  77.000000   
63       30.775000        NaN     78.000000        NaN   
64       25.000000        NaN     78.000000        NaN   
65       31.600000        NaN     78.666667        NaN   
66       27.200000  33.550000     81.000000  91.000000   
67       23.900000  38.500000     70.000000  78.000000   
68       35.600000        NaN     82.000000        NaN   
69       29.628732        NaN     81.000000        NaN   
70             NaN  32.500000           NaN  82.000000   
72       19.600000        NaN     72.405184        NaN   
81       25.900000        NaN     74.000000        NaN   

        DiabetesPedigreeFunction               Glucose              \
Outcome                        0         1           0           1   
Age                                                                  
21                      0.415828  0.642600  107.735979  139.400000   
22                      0.389574  0.658273  103.973336  153.818182   
23                      0.431581  0.469571  109.032258  122.857143   
24                      0.389737  0.411750  113.131579  140.500000   
25                      0.483294  0.885143   95.382353  145.785714   
26                      0.413640  0.412875  114.080000  131.125000   
27                      0.430167  0.596500  110.708333  129.000000   
28                      0.437320  0.515400  109.800000  145.200000   
29                      0.373750  0.452154  114.187500